util/promlint/promlint.go

```diff
@@ -0,0 +1,187 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package promlint provides a linter for Prometheus metrics.
+package promlint
+
+import (
+	"fmt"
+	"io"
+	"sort"
+	"strings"
+
+	dto "github.com/prometheus/client_model/go"
+	"github.com/prometheus/common/expfmt"
+)
+
+// A Linter is a Prometheus metrics linter.  It identifies issues with metric
+// names, types, and metadata, and reports them to the caller.
+type Linter struct {
+	r io.Reader
+}
+
+// A Problem is an issue detected by a Linter.
+type Problem struct {
+	// The name of the metric indicated by this Problem.
+	Metric string
+
+	// A description of the issue for this Problem.
+	Text string
+}
+
+// New creates a new Linter that reads an input stream of Prometheus metrics.
+// Only the text exposition format is supported.
+func New(r io.Reader) *Linter {
+	return &Linter{
+		r: r,
+	}
+}
+
+// Lint performs a linting pass, returning a slice of Problems indicating any
+// issues found in the metrics stream.
+func (l *Linter) Lint() ([]Problem, error) {
+	// TODO(mdlayher): support for protobuf exposition format?
+	d := expfmt.NewDecoder(l.r, expfmt.FmtText)
+
+	var problems []Problem
+
+	var mf dto.MetricFamily
+	for {
+		if err := d.Decode(&mf); err != nil {
+			if err == io.EOF {
+				break
+			}
+
+			return nil, err
+		}
+
+		problems = append(problems, lint(mf)...)
+	}
+
+	// Ensure deterministic output.
+	sort.SliceStable(problems, func(i, j int) bool {
+		if problems[i].Metric < problems[j].Metric {
+			return true
+		}
+
+		return problems[i].Text < problems[j].Text
+	})
+
+	return problems, nil
+}
+
+// lint is the entry point for linting a single metric.
+func lint(mf dto.MetricFamily) []Problem {
+	// TODO(mdlayher): lint rules for specific metrics types.
+	return lintGeneral(mf)
+}
+
+// lintGeneral detects general issues not tied to any specific metric type.
+func lintGeneral(mf dto.MetricFamily) []Problem {
+	var problems []Problem
+
+	// Expect all metrics to have help text available.
+	if mf.Help == nil {
+		problems = append(problems, Problem{
+			Metric: *mf.Name,
+			Text:   "no help text",
+		})
+	}
+
+	problems = append(problems, lintMetricUnits(mf)...)
+
+	return problems
+}
+
+// lintMetricUnits detects issues with metric unit names.
+func lintMetricUnits(mf dto.MetricFamily) []Problem {
+	var problems []Problem
+
+	// Expect all metric names to use base units.
+	for _, u := range baseUnits {
+		for _, p := range unitPrefixes {
+			// Assume that metrics names will be of the form "foo_units".
+			unit := p + u
+			if !strings.Contains(*mf.Name, "_"+unit) {
+				continue
+			}
+
+			problems = append(problems, Problem{
+				Metric: *mf.Name,
+				Text:   fmt.Sprintf("use base unit %q instead of %q", u, unit),
+			})
+
+			// Special case: enforce the international spelling "metres" over the
+			// American spelling "meters".
+			if strings.Contains(*mf.Name, "meters") {
```

Comment:
What is 'international spelling'? Isn't this simply british english? I also find it as 'metres (international spelling)' on wikipedia, but the links sends me to https://en.wikipedia.org/wiki/American_and_British_English_spelling_differences#-re.2C_-er

I personally prefer meters. I like how the english language envolves in the US and consider -re an artifact kept by conservationist artifically alive and I don't want to become accomplice in that :).

Realistically I wouldn't vet for something like that.